firedrake/slate/slac/kernel_builder.py

```diff
@@ -106,102 +186,50 @@ def prepare_coefficients(expression):
+    where `terminal_node` objects are :class:`slate.Tensor` nodes, and
+    `symbol_name` are :class:`coffee.base.Symbol` objects. In addition,
+    this function will return a list `aux_exprs` of any expressions that
+    require special handling in the compiler. This includes expressions
+    that require performing operations on already assembled data.
 
-    This mapping is used in the :class:`SlateKernelBuilder` to provide direct access to all
-    temporaries associated with a particular slate expression.
+    This mapping is used in the :class:`KernelBuilder` to provide direct
+    access to all temporaries associated with a particular slate expression.
 
     :arg expression: a :class:`slate.TensorBase` object.
-    :arg temps: a dictionary that becomes populated recursively and is later returned as the
-                temporaries map. This argument is initialized as an empty `dict` before
-                recursion starts.
-    :arg aux_exprs: a list that becomes populated recursively and is later returned as the
-                    list of auxiliary expressions that require special handling in SLATE's
-                    linear algebra compiler
+    :arg temps: a dictionary that becomes populated recursively and is later
+                returned as the temporaries map. This argument is initialized
+                as an empty `dict` before recursion starts.
+    :arg aux_exprs: a list that becomes populated recursively and is later
+                    returned as the list of auxiliary expressions that require
+                    special handling in Slate's linear algebra compiler
 
     Returns: the arguments temps and aux_exprs.
     """
     # Prepare temporaries map and auxiliary expressions list
     if temps is None:
-        temps = {}
-    else:
-        temps = temps
+        temps = collections.defaultdict()
 
     if aux_exprs is None:
         aux_exprs = []
-    else:
-        aux_exprs = aux_exprs
 
-    if isinstance(expression, Tensor):
-        if expression not in temps.keys():
-            temps[expression] = ast.Symbol("T%d" % len(temps))
+    if isinstance(expr, Tensor):
+        if expr not in temps:
+            temps[expr] = ast.Symbol("T%d" % len(temps))
 
-    elif isinstance(expression, Action):
-        # This is a special case where we need to handle this expression separately from the rest
-        aux_exprs.append(expression)
+    elif isinstance(expr, Action):
+        aux_exprs.append(expr)
         # Pass in the acting tensor to extract any necessary temporaries
-        prepare_temps_and_aux_exprs(expression.tensor, temps=temps, aux_exprs=aux_exprs)
-
-    elif isinstance(expression, (UnaryOp, BinaryOp)):
-        for operand in expression.operands:
-            prepare_temps_and_aux_exprs(operand, temps=temps, aux_exprs=aux_exprs)
+        generate_expr_data(expr.operands[0], temps=temps, aux_exprs=aux_exprs)
 
+    elif isinstance(expr, (UnaryOp, BinaryOp)):
```

Comment:
This could have just been `isinstance(expr, TensorOp)` now, I think.  But doesn't matter.